library/src/main/java/com/google/android/exoplayer/hls/parser/MpaReader.java

```diff
@@ -0,0 +1,297 @@
+                    // Layer1
+                    {  0,  32,  48,  56,  64,  80,  96, 112, 128, 144, 160, 176, 192, 224, 256},
+                    // Layer2
+                    {  0,   8,  16,  24,  32,  40,  48,  56,  64,  80,  96, 112, 128, 144, 160},
+                    // Layer3
+                    {  0,   8,  16,  24,  32,  40,  48,  56,  64,  80,  96, 112, 128, 144, 160}
+            }
+    };
+
+    /**
+     * Samples per Frame:
+     *
+     *  @index LSF
+     *  @index Layer
+     */
+
+    private static final int[][] MPA_SAMPLES_PER_FRAME = new int[][] {
+            {           // MPEG 1
+                    384,   // Layer1
+                    1152,   // Layer2
+                    1152    // Layer3
+            },
+            {           // MPEG 2, 2.5
+                    384,   // Layer1
+                    1152,   // Layer2
+                    576    // Layer3
+            }
+    };
+
+    /**
+     * Coefficients (samples per frame / 8):
+     *
+     * @index = LSF
+     * @index = Layer
+     */
+
+    private static final int[][] MPA_COEFFICIENTS = new int[][] {
+            {           // MPEG 1
+                    12,    // Layer1
+                    144,    // Layer2
+                    144     // Layer3
+            },
+            {           // MPEG 2, 2.5
+                    12,    // Layer1
+                    144,    // Layer2
+                    72     // Layer3
+            }
+    };
+
+    /**
+     * slot size per layer:
+     *
+     * @index = Layer
+     */
+
+    private static final int[] MPA_SLOT_SIZE = new int[] {
+            4,          // Layer1
+            1,          // Layer2
+            1           // Layer3
+    };
+
+    public MpaReader(BufferPool bufferPool) {
+        super(bufferPool);
+        mpaScratch = new ParsableBitArray(new byte[HEADER_SIZE + CRC_SIZE]);
+        state = STATE_FINDING_SYNC;
+    }
+
+    @Override
+    public void consume(ParsableByteArray data, long pesTimeUs, boolean startOfPacket) {
+        if (startOfPacket) {
+            timeUs = pesTimeUs;
+        }
+        while (data.bytesLeft() > 0) {
+            switch (state) {
+                case STATE_FINDING_SYNC:
+                    if (skipToNextSync(data)) {
+                        bytesRead = 0;
+                        state = STATE_READING_HEADER;
+                    }
+                    break;
+                case STATE_READING_HEADER:
+                    int targetLength = hasCrc ? HEADER_SIZE + CRC_SIZE : HEADER_SIZE;
+                    if (continueRead(data, mpaScratch.getData(), targetLength)) {
+                        parseHeader();
+                        startSample(timeUs);
+                        /**
+                         * Reset buffer after Header data extraction in order to pass full
+                         *  mpeg audio frame to the mpeg decoder
+                         */
+                        data.setPosition(data.getPosition() - targetLength);
```

Comment:
The mpeg decoder needs the complete frame including the header to decode de mpeg audio, in the case of AAC it is wrapped in an adts wrapper that is what its removed